relay-quotas/src/redis.rs

```diff
@@ -125,6 +137,189 @@ impl std::ops::Deref for RedisQuota<'_> {
     }
 }
 
+/// Possible outcomes from attemping to decrement a global counter budget.
+#[derive(Debug)]
+enum DecOutcome {
+    UpstreamExceeded,
+    NoRateLimit,
+    KeyMissing,
+    PoisonedLock,
+    SlotExpired,
+    BudgetEmpty,
+}
+
+#[derive(Debug, thiserror::Error)]
+enum GlobalRateLimitError {
+    #[error("quota key missing")]
+    KeyMissing,
+    #[error("failed to access RWLock")]
+    PoisonedLock,
+    #[error("slot expired")]
+    SlotExpired,
+    #[error("budget empty")]
+    BudgetEmpty,
+    #[error("failed to communicate with redis")]
+    Redis,
+}
+
+/// Counters used as a cache for global quotas.
+///
+/// We might have different global counters for different conditions (e.g. different namespaces).
+/// We use an RwLock because the counter itself is an `AtomicUsize`. Atomics can be modified with
+/// a shared reference, meaning that we can have multiple threads updating the counter concurrently,
+/// as long as there's no writers.
+#[derive(Clone, Default)]
+pub struct GlobalCounters(Arc<RwLock<hashbrown::HashMap<BudgetKey, BudgetState>>>);
+
+impl GlobalCounters {
+    fn update_limit(
+        &self,
+        budget_key: BudgetKey,
+        current_slot: u64,
+        new_budget: usize,
+        redis_value: u64,
+    ) -> Result<(), GlobalRateLimitError> {
+        let mut map = self
+            .0
+            .write()
+            .map_err(|_| GlobalRateLimitError::PoisonedLock)?;
+
+        // If there's still some budget left in the same slot we will add it to the new budget we took.
+        // This can happen if for example the budget is 30, then we try to decrement by 50, so we
+        // ask for 100 more in budget but the 30 budget is still available so the local count is 130.
+        let old_budget: usize = map
+            .get(&budget_key)
+            .filter(|val| val.slot == current_slot)
+            .map(|val| val.count.load(Ordering::SeqCst))
```

Comment:
Why do we have both atomic counters _and_ a `RwLock`? Wouldn't the outer lock be enough to protect the data?

Reply:
If we didn't have atomics within the lock, we would need write access to modify the budget, which would slow down multithreading a lot. For now, we only need write access on the lock when calling redis_sync. Now that i think of it, I could optimize it further by making `slot` an atomic, and only getting write access when we insert a new key.

Reply:
^ implemented said optimization